webapp/loggedin/modules/data/dataVis/dataQuery/components/tableColumnEdit.js

```diff
@@ -1,77 +1,67 @@
+import { useDispatch } from 'react-redux'
+import PropTypes from 'prop-types'
 
+import * as NodeDef from '@core/survey/nodeDef'
 import * as NodeDefLayout from '@core/survey/nodeDefLayout'
 import * as Record from '@core/record/record'
 import * as Node from '@core/record/node'
-import * as Authorizer from '@core/auth/authorizer'
 
-import * as SurveyState from '@webapp/survey/surveyState'
-import * as AppState from '@webapp/app/appState'
-import NodeDefTableCellBody from '../../../../../surveyViews/surveyForm/nodeDefs/components/nodeDefTableCellBody'
+import { useAuthCanEditRecord, useSurveyCycleKey, useSurveyInfo } from '@webapp/commonComponents/hooks'
 
-import { createNodePlaceholder, removeNode, updateNode } from '../../../../../surveyViews/record/actions'
+import NodeDefTableCellBody from '@webapp/loggedin/surveyViews/surveyForm/nodeDefs/components/nodeDefTableCellBody'
 
-class TableColumnEdit extends React.Component {
-  render() {
-    const {
-      surveyInfo,
-      surveyCycleKey,
-      canEditRecord,
-      nodeDef,
-      record,
-      cell,
-      updateNode,
-      removeNode,
-      createNodePlaceholder,
-    } = this.props
+import { createNodePlaceholder, removeNode, updateNode } from '@webapp/loggedin/surveyViews/record/actions'
 
-    if (cell) {
-      const { parentUuid, node } = cell
+const TableColumnEdit = (props) => {
+  const { nodeDef, record, cell } = props
 
-      const parentNode = {
-        [Node.keys.recordUuid]: Record.getUuid(record),
-        [Node.keys.parentUuid]: parentUuid,
-      }
+  if (cell) {
+    const { parentUuid, node } = cell
 
-      return (
-        <NodeDefTableCellBody
-          surveyInfo={surveyInfo}
-          surveyCycleKey={surveyCycleKey}
-          nodeDef={nodeDef}
-          parentNode={parentNode}
-          nodes={[node]}
-          entry={true}
-          entryDataQuery={true}
-          edit={false}
-          renderType={NodeDefLayout.renderType.tableBody}
-          canEditRecord={canEditRecord}
-          updateNode={updateNode}
-          removeNode={removeNode}
-          createNodePlaceholder={createNodePlaceholder}
-          windowed={false}
-        />
-      )
+    const surveyInfo = useSurveyInfo()
+    const surveyCycleKey = useSurveyCycleKey()
+    const canEditRecord = useAuthCanEditRecord(record)
+    const dispatch = useDispatch()
+
+    const parentNode = {
+      [Node.keys.recordUuid]: Record.getUuid(record),
+      [Node.keys.parentUuid]: parentUuid,
     }
 
-    return null
+    return (
+      <NodeDefTableCellBody
+        className="test"
+        surveyInfo={surveyInfo}
+        surveyCycleKey={surveyCycleKey}
+        nodeDef={nodeDef}
+        parentNode={parentNode}
+        nodes={[node]}
+        entry
+        entryDataQuery
+        edit={false}
+        renderType={NodeDefLayout.renderType.tableBody}
+        canEditRecord={canEditRecord}
+        readOnly={NodeDef.isReadOnly(nodeDef) || NodeDef.isAnalysis(nodeDef)}
```

Comment:
can readOnly be calculated inside NodeDefTableCellBody? it has the nodeDef

Reply:
yes... I didn't want to change even NodeDefTableCellBody...

Reply:
but doest NodeDefTableBody needs readOnly prop?

Reply:
that prop is passed to the NodeDef UI component together with all the props